### PR TITLE
Implementing node-glob's `ignore` option, fixes #24

### LINKED ISSUE #24: Using `src` globs can be really slow with negation
Trying to use negation in globs can be extremely slow if there are many files in the negated paths (e.g., `node_modules`). For example:

```
[ '**/*.js', '!node_modules/**' ]
```

We're still in the initial phases of a project and while that pattern works, it was causing a lint task to take more than **4s** to finish. Changing the pattern to only use positives (but that means manually adding half a dozen paths and files) brought the task time down to 200ms.

### PR DESCRIPTION
**Pull request by UltCombo**
Quite a few tests are broken on my machine, not sure if I'm missing something or the `ignore` option is buggy in regard to the `dot` option and absolute paths (on Windows).

Fixes #24

### CODE REVIEW

**Comment by contra** [comment]
Actually if it cleans up the code lets go ahead and remove regexp support, I've never seen anyone use it. No need to keep that feature.

**Comment by UltCombo** [comment]
Something odd is that Mocha was seemingly "cascading" certain test fails -- e.g. in my first commit above, the "glob-stream create() should handle RegExps as negative matchers" test shows up as a fail, but it passes when run in isolation.

File: index.js [hunk 1]
```diff
@@ -17,6 +17,13 @@ var gs = {
     // remove path relativity to make globs make sense
     ourGlob = unrelative(opt.cwd, ourGlob);
 
+    if (negatives.length) {
+      // Do not mutate the options object which may be used
+      // in multiple `gs.createStream()` calls.
+      opt = assign({}, opt);
```
Comment (contra) [inline index.js]: wouldnt a lodash.clone be better here?

**Comment by UltCombo** [comment]
I've removed RegExp support, but no change in the failed tests. I wonder if it is an issue with glob's `ignore` option, or if I'm just too sleepy to see an obvious error.

File: index.js [hunk 2]
```diff
@@ -67,20 +65,15 @@ var gs = {
...
 
-      // create Minimatch instances for negative glob patterns
-      if (globArray === negatives && typeof glob === 'string') {
-        glob = new Minimatch(unrelative(opt.cwd, glob), opt);
-      }
-
       globArray.push({
         index: index,
-        glob: glob
+        glob: globArray === negatives ? glob.slice(1) : glob
```
Comment (contra) [inline index.js]: should just be `isNegative(glob) ? glob.slice(1) : glob` then ditch the globArray var

File: index.js [hunk 3]
```diff
@@ -105,11 +98,6 @@ var gs = {
...
 };
 
-function isMatch(file, matcher) {
-  if (matcher instanceof Minimatch) return matcher.match(file.path);
-  if (matcher instanceof RegExp) return matcher.test(file.path);
-}
-
 function isNegative(pattern) {
   if (typeof pattern === 'string') return pattern[0] === '!';
   if (pattern instanceof RegExp) return true;
```
Comment (contra) [inline index.js]: remove this line since we are dropping regexps

File: index.js (hunk 1 above)
Comment (UltCombo) [inline index.js]: I tend to use `object-assign` as its logic is rather simple and well-spec'd (follows ECMAScript 2015 spec.). But yes, I guess a clone or cloneDeep would suffice as well.

Comment (UltCombo) [inline index.js]: Should I add the entire lodash lib as a dependency or just the clone method?

**Comment by UltCombo** [comment]
Oh -- I'm forgetting to `unrelative` the negative globs, doh.

File: index.js (hunk 3 above)
Comment (UltCombo) [inline index.js]: done

File: index.js (hunk 2 above)
Comment (UltCombo) [inline index.js]: what about the `globArray.push` above? Should I inline `(isNegative(glob) ? glob.slice(1) : glob).push()` as well?

File: index.js (hunk 1 above)
Comment (UltCombo) [inline index.js]: done

**Comment by UltCombo** [comment]
Looks like there is still an issue with unrelative logic.

File: index.js (hunk 2 above)
Comment (contra) [inline index.js]: i would put isNegative in a var so it only gets called once then ternary the other two places

**Comment by UltCombo** [comment]
Oh wait, the unrelative logic is fine -- node-glob's ignore option seems to be screwing it up independent of using relative or absolute paths.

File: index.js (hunk 2 above)
Comment (UltCombo) [inline index.js]: done

**Comment by UltCombo** [comment]
"should return a file name stream with dotfiles negated" is failing due to a [node-glob issue](https://github.com/isaacs/node-glob/issues/166).

**Comment by UltCombo** [comment]
I thought the error could be due to passing absolute paths (which use `\` as dir separator in Windows) to the `ignore` option (according to the [docs](https://github.com/isaacs/node-glob#windows), node-glob always interprets `\` as escape sequences).

Minimatch@2 does treat `\` as a valid dir separator, but node-glob currently does not.

The test is still failing on Travis (Linux), so there must be yet another error...

**Comment by contra** [comment]
Any progress on this?

**Comment by UltCombo** [comment]
@contra there are some blocking issues:

- Awaiting @isaacs's feedback on https://github.com/isaacs/node-glob/pull/166#issuecomment-76290480
- It also seems like node-glob was not supposed to work with absolute paths like glob-stream does currently (mainly, absolute Windows paths). I tried rewriting the logic to resolve the paths after the glob matching step, but that also broke quite a few tests.

**Comment by UltCombo** [comment]
@contra If you don't mind the breaking change regarding the `dot` behavior, we can skip those tests temporarily and I can try fixing the remaining logic again.

**Comment by contra** [comment]
@UltCombo Seems like the resolution from the glob issue was the ball is in your court. Do you need any help with this?

https://github.com/isaacs/node-glob/pull/166#issuecomment-76292453

**Comment by UltCombo** [comment]
@contra Well yes, I have near to no experience with the node-glob code base, and a good part of it seems nearly unreadable to me.

**Comment by thirdcreed** [comment]
Am I wrong in thinking that it's impossible to have regex support pre-read?  Because you have to have access to the file names to test against?

**Comment by thirdcreed** [comment]
I nearly have a working PR for this, there's one test broken: 
'should return a file name stream with dotfiles negated'
and it's timing out at 2000ms.

Any thoughts on what that might be, for some reason the stream is not emitting anything.
https://github.com/thirdcreed/glob-stream/blob/master/index.js

**Comment by thirdcreed** [comment]
https://github.com/isaacs/node-glob/issues/232  This is a block for me.  I think there's some bigger problem involving globs that end with "/**", a completely different matcher is used for those cases.  There's some kind of race condition.  It emits before the on end event is registered.

**Comment by thirdcreed** [comment]
I'm going to wait until issacs has time to answer it, I guess this above my head. Apparently the race condition thing is understood, It looks like its on the user to guarantee their event.   Unless issacs has some insight on it, i'm out of ideas, and I can't seem to get that one fucking test to run. Ah well.  It looks like @UltCombo's on this again anyway.

**Comment by UltCombo** [comment]
There are 23 failing tests (on both current master and @thirdcreed's branch) when testing on Windows. This does not make it easy to work on a fix and I'm relatively short on free time.

I guess I can give this another shot on my Linux or Mac box soon.

@contra May I suggest adding [AppVeyor](http://www.appveyor.com/) (Windows CI)? Here's a sample [appveyor.yml](https://github.com/floatdrop/gulp-watch/blob/master/appveyor.yml) just in case.

**Comment by contra** [comment]
@UltCombo We use appveyor on some other stuff, feel free to add it if you need it

**Comment by UltCombo** [comment]
@contra Do you have an appveyor config of your choice to add to this repository? Or should I just create one using the same Node.js versions as this repo's Travis config?

**Comment by phated** [comment]
https://github.com/gulpjs/gulp-cli/blob/master/appveyor.yml

**Comment by jonschlinkert** [comment]
> Something odd is that Mocha was seemingly "cascading" certain test fails

if there are multiple tests with the same description, mocha will run them all, even if one has `.only` on it. I remember this driving me crazy a few times before @doowb noticed what was happening and pointed it out. again, not sure if it's related. but if not, maybe it will save someone else from going crazy lol.

**Comment by phated** [comment]
Closing in favor of #71 which is passing all the tests.  Thanks for putting in a lot of the work here @UltCombo

**Comment by UltCombo** [comment]
Oh, this was still on my TODO list, but kept falling off the radar.

No problem, and indeed, #71 seems more suitable and up-to-date. I'll comment there if I find anything odd.

**Pull request closed without merging**